aea/cli/publish.py

```diff
@@ -139,33 +152,185 @@ def _check_is_item_in_local_registry(
         )
     except click.ClickException as e:
         raise click.ClickException(
-            f"Dependency is missing. {str(e)}\nPlease push it first and then retry."
+            f"Dependency is missing. {str(e)}\nPlease push it first and then retry or use {PUSH_ITEMS_FLAG} flag to push automatically"
         )
 
 
-def _save_agent_locally(ctx: Context, is_mixed: bool = False) -> None:
+class BaseRegistry(ABC):
+    """Base registry class."""
+
+    @abstractmethod
+    def check_item_present(self, item_type_plural: str, public_id: PublicId) -> None:
+        """
+        Check item present in registry.
+
+        Raise ClickException if not found.
+
+        :param item_type_plural: str, item type.
+        :param public_id: PublicId of the item to check.
+
+        :return: None
+        """
+
+    @abstractmethod
+    def push_item(self, item_type_plural: str, public_id: PublicId) -> None:
+        """
+        Push item to registry.
+
+        :param item_type_plural: str, item type.
+        :param public_id: PublicId of the item to check.
+
+        :return: None
+        """
+
+    def check_item_present_and_push(
+        self, item_type_plural: str, public_id: PublicId
+    ) -> None:
+        """
+        Check item present in registry and push if needed.
+
+        Raise ClickException if not found.
+
+        :param item_type_plural: str, item type.
+        :param public_id: PublicId of the item to check.
+
+        :return: None
+        """
+
+        with suppress(click.ClickException):
+            return self.check_item_present(item_type_plural, public_id)
+
+        try:
+            self.push_item(item_type_plural, public_id)
+        except Exception as e:
+            raise click.ClickException(
+                f"Failed to push missing item: {item_type_plural} {public_id}: {e}"
+            ) from e
+
+        try:
+            self.check_item_present(item_type_plural, public_id)
+        except Exception as e:
+            raise click.ClickException(
+                f"Failed to find item after push: {item_type_plural} {public_id}: {e}"
+            ) from e
+
+
+class LocalRegistry(BaseRegistry):
```

Comment:
Might as well have a mixed case?

Reply:
In my opinion, the mixed case should be handled by a separate registry class, which is a composition of `LocalRegistry` and `RemoteRegistry`.

Reply:
But for the scope of this PR it is probably OK to have the mixed mode implemented in `LocalRegistry`.

Reply:
updated